File: 3. Algoritmos básicos/JavaScript/busqueda.js
// 1. Escribe una función que realice una búsqueda secuencial en un arreglo y devuelva el índice del elemento encontrado.
// 2. Implementa la búsqueda binaria en un arreglo ordenado de números.

// Búsqueda secuencial
function busquedaSecuencial(arr, elemento) {
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === elemento) {
      return i;
    }
  }
  return -1;
}

// Búsqueda binaria
function busquedaBinaria(arr, elemento) {
  let inicio = 0;
  let fin = arr.length - 1;
  while (inicio <= fin) {
    let medio = Math.floor((inicio + fin) / 2);
    if (arr[medio] === elemento) {
      return medio;
    } else if (arr[medio] < elemento) {
      inicio = medio + 1;
    } else {
      fin = medio - 1;
    }
  }
  return -1;
}

let arr = [1, 2, 3, 4, 5, 6, 7, 8, 9];
let numero = 5;
// Prueba de la búsqueda secuencial
console.log(busquedaSecuencial(arr, numero)); // 4

// Prueba de la búsqueda binaria
console.log(busquedaBinaria(arr, numero)); // 4

File: 3. Algoritmos básicos/Python/busqueda.py
# 1. Implementa una búsqueda secuencial para encontrar un número en una lista y devuelve su posición.
# 2. Crea una función que implemente la búsqueda binaria en una lista ordenada.


def busqueda_secuencial(lista, numero):
    for i in range(len(lista)):
        if lista[i] == numero:
            return i
    return -1


def busqueda_binaria(lista, numero):
    izquierda = 0
    derecha = len(lista) - 1
    while izquierda <= derecha:
        mitad = (izquierda + derecha) // 2
        if lista[mitad] == numero:
            return mitad
        elif lista[mitad] < numero:
            izquierda = mitad + 1
        else:
            derecha = mitad - 1
    return -1


lista = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
numero = 13

# Prueba de la búsqueda secuencial
print(busqueda_secuencial(lista, numero))  # 6

# Prueba de la búsqueda binaria
print(busqueda_binaria(lista, numero))  # 6
